### PR TITLE
fix(server): XEP-0198 resume loses stored presence payloads (#1206)

### PR DESCRIPTION
**Pull request by randax**
Fixes #1206.

Follow-up to #1101 (PR #1205) and #1103 (PR #1207), found in adversarial review.

## Problem

PR #1207 wired a resource's presence extension payloads (XEP-0115 `<c/>`,
XEP-0319 `<idle/>`, arbitrary extensions) through the **in-memory** XEP-0198
detached-session boundary: `DetachedSession.presence_payloads` is carried into
resume (`registration.rs`) and served from the detached-window probe /
subscription-delivery paths.

The **durable** persistence layer was explicitly deferred (a `Vec::new()`
placeholder in `persisted_to_detached`). `PersistedSession` had no payloads
field, so `detached_to_persisted` dropped them on write and
`persisted_to_detached` hard-coded empty on read. Every session rehydrated
from durable storage — a **cross-node resume** (clustered) or a **restart
resume** (single-node) — therefore came back caps-less. Since XEP-0198 resume
means the client does **not** resend presence, every subsequent probe response
and subscription-approval push then relayed that resource as available with no
`<c/>` for the rest of the session, breaking feature detection (e.g. Jingle)
toward its subscribers — the same class of bug #1101 fixed, reintroduced
through the durable-rehydration door. (RFC 6121 §4.3.2 requires a probe
response to reproduce the complete presence stanza.)

## Fix

Persist the payloads in the SM durable shape and restore them:

- `PersistedSession` gains `presence_payloads: Vec<minidom::Element>` — typed
  arbitrary-XML per the typed-payloads hard rule. The two codec conversions
  (`detached_to_persisted` / `persisted_to_detached`) carry the field.
- New `sm_sessions.presence_payloads` TEXT column (portable **and**
  Postgres-fenced backends). Encoded as one `minidom` wrapper element built via
  the element builder (never `format!`/string concat, per the XML hard rule);
  serialization to the column happens only at the DB I/O boundary, parsed back
  to typed exactly once on read. Empty payloads → NULL column → empty vec.
- The shared `decode_session` now reads column 17, so the plain SELECTs and the
  `list_all_sessions_with_unacked` JOIN both carry it; the JOIN's unacked reads
  shift to 18/19/20 (existing JOIN/poison-pill tests guard the shift).
- The wrapper's `urn:waddle:sm:presence-payloads:0` namespace is an internal
  storage encoding only — never on the wire; only the wrapper's children (the
  original extension elements) are ever relayed.
- `decode_session` degrades a malformed `presence_payloads` cell to caps-less
  (log + empty) rather than failing the whole decode: caps are non-essential
  and self-heal on the next presence broadcast, whereas poisoning the session
  would drop its XEP-0198 unacked message queue on cold start — a strictly
  worse trade (review hardening).

Out of scope (noted in #1206): `user_actor.rs` `UpdatePresenceState` still
stores empty payloads (no production senders today); ADR-0017 later phases that
cut probe responses over to the actor store must carry payloads too.

## Tests (TDD, vertical slices — RED→GREEN per slice)

- `xep0198_persistence_round_trip_preserves_presence_payloads` (XEP-0198 suite):
  store a detached session with real XEP-0115 caps + XEP-0319 idle, persist,
  restart the registry, restore — a probe of the still-detached resource
  (`detached_presence_state`) returns the payloads verbatim (AC1 + AC2).
- `round_trip_session_preserves_presence_payloads` (portable SQLite backend):
  `upsert_session` → `get_session` preserves payloads.
- `store_session_atomic_and_join_read_preserve_presence_payloads`: the real
  production store path + cold-start JOIN restore preserves payloads AND the
  unacked stanza still decodes (end-to-end JOIN column-index guard).
- `presence_payloads_tests` (codec): empty→NULL→empty, in-order multi-payload
  round-trip, and malformed column → typed error (poison-pill, not a panic).
- `upsert_and_get_session_preserves_presence_payloads` (Postgres-fenced,
  env-gated) mirrors the portable round-trip.

## Acceptance criteria

- [x] After detach + resume (no presence resend), a probe for the resumed
      resource returns its pre-detach caps and extensions
- [x] Probe of a *detached* (not yet resumed) available resource carries its
      stored payloads
- [x] Dedicated Rust test in the XEP-0198/XEP-0115 suites

## Verification

`cargo fmt --check` clean; `cargo clippy -D warnings` clean for default **and**
`--features clustering`; portable backend + XEP-0198 suites green; fenced
backend compiles under `--features clustering`. Reviewed by two independent
adversarial personas (opus) plus gpt-5.5 (codex) — no real defects; the two
review-response items (stale doc + codec edge-case tests) are folded in.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: server/crates/waddle-server/src/sm_persistence/codec.rs [hunk 1]
```diff
@@ -182,6 +207,58 @@ pub(crate) fn serialize_stanza(stanza: &Stanza) -> Result<String, SmPersistenceE
...
+pub(crate) fn parse_presence_payloads(
+    raw: Option<String>,
+) -> Result<Vec<xmpp_parsers::minidom::Element>, SmPersistenceError> {
+    let Some(raw) = raw.filter(|s| !s.is_empty()) else {
+        return Ok(Vec::new());
+    };
+    let wrapper: xmpp_parsers::minidom::Element = raw
+        .parse()
+        .map_err(|e: xmpp_parsers::minidom::Error| SmPersistenceError::Other(e.to_string()))?;
+    Ok(wrapper.children().cloned().collect())
```
[the remark this answers is not in the document]
Comment (randax) [reply server/crates/waddle-server/src/sm_persistence/codec.rs:272]: Fixed in 5e906f98. `parse_presence_payloads` now validates the root is the server-written wrapper (`payloads` in `urn:waddle:sm:presence-payloads:0`) and returns a typed error on mismatch, so a well-formed-but-wrong-root cell routes through `decode_session`'s warn+degrade-to-caps-less path instead of silently returning wrong/empty payloads. Added `well_formed_wrong_root_is_a_typed_error_so_the_caller_can_degrade`.

File: server/crates/waddle-server/src/sm_persistence/codec.rs [hunk 2]
```diff
@@ -182,6 +207,58 @@ pub(crate) fn serialize_stanza(stanza: &Stanza) -> Result<String, SmPersistenceE
...
+}
```
[the remark this answers is not in the document]
Comment (randax) [reply server/crates/waddle-server/src/sm_persistence/codec.rs:273]: Fixed in 5e906f98. `parse_presence_payloads` now validates the root is the server-written wrapper (`payloads` in `urn:waddle:sm:presence-payloads:0`) and returns a typed error on mismatch, so a well-formed-but-wrong-root cell routes through `decode_session`'s warn+degrade-to-caps-less path instead of silently returning wrong/empty payloads. Added `well_formed_wrong_root_is_a_typed_error_so_the_caller_can_degrade`.

File: server/crates/waddle-server/src/sm_persistence/codec.rs [hunk 3]
```diff
@@ -198,3 +275,45 @@ fn parse_stanza(element: xmpp_parsers::minidom::Element) -> Result<Stanza, SmPer
...
+            .unwrap();
+        let encoded = serialize_presence_payloads(&[caps.clone(), idle.clone()])
+            .unwrap()
+            .expect("non-empty payloads serialize to Some");
+        let decoded = parse_presence_payloads(Some(encoded)).unwrap();
+        assert_eq!(decoded, vec![caps, idle]);
+    }
+
+    #[test]
+    fn malformed_column_is_a_typed_error_not_a_panic() {
```
[the remark this answers is not in the document]
Comment (randax) [reply server/crates/waddle-server/src/sm_persistence/codec.rs:326]: Fixed in 5e906f98 — corrected the stale "poison pill" wording; `decode_session` now degrades a malformed cell to caps-less (preserving the unacked queue) rather than dropping the session.
